src/validate.rs

```diff
@@ -1,33 +1,1490 @@
+                    critical: true,
+                    position: None,
+                    message: String::from("No cases allowed."),
+                    suggestion: None,
+                }],
+                normalized: None,
+            };
+        } else if !config.cases.contains(&case) {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: None,
+                    message: format!("Unknown case '{}'.", case),
+                    suggestion: Some(format!("Known cases are: '{}'", config.cases.join("', '"))),
+                }],
+                normalized: None,
+            };
+        }
+    }
+    let mut translation = match ParsedString::parse(&translation) {
+        Err(err) => {
+            return ValidationResult {
+                errors: vec![ValidationError {
+                    critical: true,
+                    position: Some(err.position),
+                    message: err.message,
+                    suggestion: None,
+                }],
+                normalized: None,
+            };
+        }
+        Ok(parsed) => parsed,
+    };
+    let errs = validate_string(&config, &translation, Some(&base));
+    if errs.iter().any(|e| e.critical) {
+        ValidationResult {
+            errors: errs,
+            normalized: None,
+        }
+    } else {
+        sanitize_whitespace(&mut translation);
+        normalize_string(&config.get_dialect(), &mut translation);
+        ValidationResult {
+            errors: errs,
+            normalized: Some(translation.compile()),
+        }
+    }
+}
+
+fn remove_ascii_ctrl(t: &mut String) {
+    *t = t.replace(|c| char::is_ascii_control(&c), " ");
+}
+
+fn remove_trailing_blanks(t: &mut String) {
+    if let Some(last) = t.rfind(|c| c != ' ') {
+        t.truncate(last + 1);
+    }
+}
+
+/// Replace all ASCII control codes with blank.
+/// Remove trailing blanks at end of each line.
+fn sanitize_whitespace(parsed: &mut ParsedString) {
+    let mut is_eol = true;
+    for i in (0..parsed.fragments.len()).rev() {
+        let mut is_nl = false;
+        match &mut parsed.fragments[i].content {
+            FragmentContent::Text(t) => {
+                remove_ascii_ctrl(t);
+                if is_eol {
+                    remove_trailing_blanks(t);
+                }
+            }
+            FragmentContent::Choice(c) => {
+                for t in &mut c.choices {
+                    remove_ascii_ctrl(t);
+                }
+            }
+            FragmentContent::Command(c) => {
+                is_nl = c.name.is_empty();
+            }
+            _ => (),
+        }
+        is_eol = is_nl;
+    }
+}
+
+struct StringSignature {
+    parameters: HashMap<usize, &'static CommandInfo<'static>>,
+    nonpositional_count: BTreeMap<String, (Occurence, usize)>,
+    // TODO track color/lineno/colorstack for positional parameters
```

Comment:
still todo?

Reply:
Yes, this is part of the "bells and whistles". Currently stashed.